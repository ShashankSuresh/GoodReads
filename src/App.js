import React, { useState } from "react";
import List from "./components/List";
import Pagination from "./components/Pagination";
import axios from "axios";
import "./App.scss";

const API_KEY = process.env.REACT_APP_GOOD_DEADS_API_KEY

const App = () => {
  const [books, setbooks] = useState([]);
  const [isloading, setLoading] = useState(false);
  const [currentPage, setCurrentPage] = useState(1);
  const [booksPerPage] = useState(10);

  const fetchBooks = async () => {
    const requestUri =
      `https://cors-anywhere.herokuapp.com/` +
      `https://www.goodreads.com/search/index.xml?key=${API_KEY}&q=${document.getElementById("search-text").value}`;

    axios.get(requestUri).then(res => {
      parseXMLResponse(res.data);
    });
  };

  const parseXMLResponse = response => {
    const parser = new DOMParser();
    const XMLResponse = parser.parseFromString(response, "application/xml");
    const XMLresults = new Array(...XMLResponse.getElementsByTagName("work"));
    const searchResults = XMLresults.map(result => XMLToJson(result));
    
    setbooks(searchResults);
    setCurrentPage(true)
  };

  const XMLToJson = XML => {
    const allNodes = new Array(...XML.children);
    const jsonResult = {};
    allNodes.forEach(node => {
      if (node.children.length) {
        jsonResult[node.nodeName] = XMLToJson(node);
      } else {
        jsonResult[node.nodeName] = node.innerHTML;
      }
    });
    return jsonResult;
  };

  const indexOfLastPost = currentPage * booksPerPage;
  const indexOfFirstPost = indexOfLastPost - booksPerPage;
  const booksList = books.slice(indexOfFirstPost, indexOfLastPost);

  const paginate = pageNumber => setCurrentPage(pageNumber);

  return (
    <div className="good-deads">
      <h1 className="text-primary">Search here...</h1>
      <input type="text" id="search-text" onChange={() => fetchBooks()} />
      <List books={booksList} loading={isloading} />
      <Pagination
        booksPerPage={booksPerPage}
        totalbooks={books.length}
        paginate={paginate}
      />
    </div>
  );
};

export default App;
