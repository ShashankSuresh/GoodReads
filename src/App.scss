.good-deads {
  margin: 0 auto;
  text-align: center;

  .list-container {
    padding-inline-start: 0px;
    list-style: none;
    display: flex;
    align-items: center;
    flex-flow: column;
    list-style: none;
    .list-item {
      margin-bottom: 20px;
      border: 1px solid black;
      p {
        width: 200px;
        margin: 0 auto;
      }
    }
  }
  .pagination {
    padding-inline-start: 0px;
    list-style: none;
    display: flex;
    justify-content: center;
    .page-item {
      margin-left: 20px;
    }
  }
}