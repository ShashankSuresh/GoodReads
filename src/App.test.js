import React from "react";
import { shallow, mount } from "enzyme";
import App from "./App";
import axios from "axios";

import mockAxios from "jest-mock-axios";

jest.mock("axios", () => ({
  get: jest.fn()
}));

axios.get.mockImplementation(() =>
  Promise.resolve({
    data: {
      posts: {
        work: {
          namespaceURI: null,
          prefix: null,
          localName: "work",
          tagName: "work",
          id: "",
          className: ""
        },
        id: "4640799",
        books_count: "748",
        ratings_count: "6471131",
        text_reviews_count: "102021",
        original_publication_year: "1997",
        original_publication_month: "6",
        original_publication_day: "26",
        average_rating: "4.47",
        best_book: {
          id: "3",
          title: "Harry Potter and the Sorcerer's Stone (Harry Potter, #1)",
          author: {
            id: "1077326",
            name: "J.K. Rowling"
          },
          image_url:
            "https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/books/1474154022l/3._SX98_.jpg",
          small_image_url:
            "https://i.gr-assets.com/images/S/compressed.photo.goodreads.com/books/1474154022l/3._SY75_.jpg"
        }
      }
    }
  })
);

const props = {
  posts: {},
  fetchBooks: jest.fn(),
  parseXMLResponse: jest.fn(),
  XMLresults: jest.fn()
};

describe("fetchData", () => {
  beforeEach(() => {
    const input = document.createElement("input");
    input.setAttribute("id", "search-text");
    document.body.appendChild(input);
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it("renders", () => {
    shallow(<App />);
  });

  it("fetchBooks", async () => {
    const wrapper = mount(<App />);
    
    wrapper
      .find(".good-deads")
      .find("#search-text")
      .instance().value = "rowling";

    wrapper
      .find(".good-deads")
      .find("#search-text")
      .simulate("change");

    axios.get.mockResolvedValue("resp");
    expect(axios.get).toHaveBeenCalledTimes(1);
    await expect(axios.get()).resolves.toBe("resp");
  });
});
