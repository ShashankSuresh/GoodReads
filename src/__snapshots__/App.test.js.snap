// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchData fetches successfully data from an API 1`] = `ReactWrapper {}`;
